### DISCUSSION TITLE
AAD OIDC keeps using implicit flow

### DISCUSSION THREAD

**Discussion by arctiqaurora** [kiali/kiali#3745] [Q&A]
I am setting up OIDC integration and Kiali keeps using implicit flow instead of authorization code flow. I have tried to modify signing_key to 16 bytes, create new kiali secret with client_secret value, but it keeps using implicit flow. Any troubleshooting steps?

**Comment by israel-hdez** [comment]
The client_secret is not relevant to have working the authorization code flow.

About the `signing_key`, make sure you are referring to this one: https://github.com/kiali/kiali-operator/blob/216a1e6fbd50d46841009d26e2580ab830687007/deploy/kiali/kiali_cr.yaml#L754. I've heard of people who is encoding twice the signing_key in base64 format. So, make sure you are not encoding it twice. If the signing key is the problem you should see the following message in the Kiali logs: `Cannot use OpenId authorization code flow because signing key is not 16, 24 nor 32 bytes long`.

If that message is absent, check your OpenId metadata to see if it is publishing that the `code` response type is supported. In your browser, go to `$your_issuer_url/.well-known/openid-configuration`. You should see a JSON document. Find the `response_types_supported` property and check that it lists `code`. If `code` is not listed, Kiali won't use the authorization code flow. See a sample of the JSON document here: https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfigurationResponse.

Also, make sure you restart the Kiali pod after doing changes to the config.

Basically, that's it. Those are the only two checks that are done in Kiali and if they are satisfied, the authorization code flow will be triggered.

**Reply by arctiqaurora** [reply]
Thanks for all the details @israel-hdez . I have checked kiali logs and do not see the error message.
```
kubectl logs kiali-56df8f659d-lhhcf -n istio-system -f
I0226 04:34:59.541064       1 kiali.go:66] Kiali: Version: v1.22.1, Commit: eb8fce019ab5b31db4a414133dc599172e9634ce
I0226 04:34:59.541361       1 kiali.go:205] Using authentication strategy [openid]
I0226 04:34:59.541401       1 kiali.go:87] Kiali: Console version: 1.22.1
I0226 04:34:59.541406       1 kiali.go:284] Updating base URL in index.html with [/kiali]
I0226 04:34:59.541706       1 kiali.go:265] Generating env.js from config
I0226 04:34:59.543784       1 server.go:67] Server endpoint will start at [:20001/kiali]
I0226 04:34:59.543793       1 server.go:68] Server endpoint will serve static content from [/opt/kiali/console]
I0226 04:34:59.543802       1 metrics_server.go:18] Starting Metrics Server on [:9090]
I0226 19:27:46.163057       1 cache.go:126] Kiali Cache is active for namespaces [.*]
``` 
I've also checked the openid-configuration as described:
```
"response_types_supported":["code","id_token","code id_token","id_token token"]
```

**Reply by israel-hdez** [reply]
> I have checked kiali logs and do not see the error message.

Ops! I forgot to say to check the logs after you try to login. That message is shown only after you try to login. Before that, it's unknown to Kiali if authentication code flow can be used.

By the way, I'm wondering how you know that `implicit flow` is being used?

**Reply by arctiqaurora** [reply]
Yes that log is what I got after trying to logging in. 
I am using Azure AAD and I am not able to get any group information in the token. only has_groups parameter is present in the token. From AAD doc it is saying that is the case for implicit flow. 
We have enabled OIDC at the cluster level and we are using RBAC at group level.

**Reply by israel-hdez** [reply]
What Kiali version are you using?

**Reply by arctiqaurora** [reply]
It is on version 1.22.1

**Reply by israel-hdez** [reply]
That explains a lot.
I'm looking at PRs, and I see that support for authorization code flow was added in version 1.24: https://github.com/kiali/kiali/pull/3189. You will need to upgrade your Kiali.

**Comment by jmazzitelli** [comment]
Release Notes for 1.24.0 - you will see auth-flow was added then. So 1.22 is a no-go if you want auth-flow:

https://kiali.io/news/release-notes/#_1_24_0

```
1.24.0
Sprint Release: September 18th, 2020
...
Features:
* OIDC - Add support for "Authentication code" flow
```

**Comment by arctiqaurora** [comment]
Thanks for the information @jmazzitelli @israel-hdez 
I have upgraded to 1.24.0. 
Now I keep seeing the error `Cannot use OpenId authorization code flow because signing key is not 16, 24 nor 32 bytes long` even when my signing key is 16 bytes. 
Do you have an example config for signing key?
I think if I leave it blank, a 16 byte key will be auto generated, but that also gave me the same signing key error.

**Reply by arctiqaurora** [reply]
actually i am seeing a 401 Unauthorized error:
```
{"error":"failure when retrieving user identity","detail":"request failed (HTTP response status = 401 Unauthorized)"}

**Reply by jmazzitelli** [reply]
> Do you have an example config for signing key?

How are you installing? Operator helm chart? Server helm chart? OLM?

I'll assume you are using the operator.... when using operator, here's how you define the signing key (by default, [a secret will be created that includes a randomly generated 16-char key](https://github.com/kiali/kiali-operator/blob/master/roles/v1.24/kiali-deploy/tasks/main.yml#L462-L481)):

https://github.com/kiali/kiali-operator/blob/v1.24/deploy/kiali/kiali_cr.yaml#L738-L754

So an example would be in the Kiali CR:

```
login_token:
  signing_key: abc1234567890def
```

> {"error":"failure when retrieving user identity","detail":"request failed (HTTP response status = 401 Unauthorized)"}

As for the above error, that is occurring [here](https://github.com/kiali/kiali/blob/v1.24/handlers/authentication.go#L672-L676). 

@israel-hdez would know more, but my first suspicion would be the configuration is incorrect, causing [this code that guesses the Kiali endpoint to fail](https://github.com/kiali/kiali/blob/v1.24.0/util/httputil/http_util.go#L90-L96). Check your [config for web_fqdn and web_schema](https://github.com/kiali/kiali-operator/blob/v1.24/deploy/kiali/kiali_cr.yaml#L791-L809), and also read [this FAQ](https://kiali.io/documentation/v1.24/faq/#x-forwarded-port).

**Reply by arctiqaurora** [reply]
I am using anthos service mesh istioctl to install, Kiali is not configured using CR, i am using configmap for all configuration. 
Install guide: https://cloud.google.com/service-mesh/docs/gke-on-prem-install

**Reply by israel-hdez** [reply]
The `request failed (HTTP response status = 401 Unauthorized)` is a sign that the authentication server is rejecting the request to fetch a token. Usually, the `unauthorized` response is because the `client_secret` is absent or is wrong. The docs have a sample command to create the required client secret. Read: https://kiali.io/documentation/latest/configuration/authentication/openid/#_set_up_with_rbac_support

After you create the `client_secret`, restart the Kiali pod and try again.

**Reply by arctiqaurora** [reply]
we've checked the secret, there's not issue with the old secret but we updated it anyways. 
we do not see the 401 error anymore from the browser, but from kaili pod log we are still seeing the same error `Not handling OpenId code flow authentication: No nonce code present. Login window timed out.` and the webpage will be redirected to /kiali with no login information.

**Reply by israel-hdez** [reply]
What message are you seeing in the browser now?
The `Not handling OpenId code flow authentication: No nonce code present. Login window timed out.` in the logs is usually an informative message (not labeled as a warning, nor error). So, it can be ignored most of the time.

The redirect with no login information is expected. That's the point of the authorization code flow (to not reveal any credentials in the URL).

**Reply by arctiqaurora** [reply]
I got redirected to the login page with url `https://xxxxxx/kiali/#`

**Reply by israel-hdez** [reply]
To further help, I would like to know what's the message shown in the browser. The URL tells me nothing.

**Reply by arctiqaurora** [reply]
that's also where i get confused as well. after clicking on `Log In With OpenID`, i am being redirected back to the same page with no message:
<img width="1436" alt="Screen Shot 2021-03-03 at 3 07 27 PM" src="https://user-images.githubusercontent.com/60113774/109865555-2cfe3b80-7c32-11eb-94df-882d63c5c43e.png">
We are exposing kiali through istio gateway:
```
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: kiali
  namespace: istio-system
spec:
  hosts:
  - "kiali.xxxxxx"
  gateways:
  - corp-ingress-gateway
  http:
  - route:
    - destination:
        host: kiali
        port:
          number: 20001

**Reply by israel-hdez** [reply]
Apparently, you are facing an old issue: https://github.com/kiali/kiali/issues/3103

**Reply by israel-hdez** [reply]
That issue was fixed some time ago. If you are already using the latest Kiali version, it's a regression.
So, if you are using latest Kiali, please open a bug report: https://github.com/kiali/kiali/issues/new/choose

If you are NOT using the latest Kiali version, please upgrade to check if the issue persists.

**Reply by arctiqaurora** [reply]
I am using version 1.30. yes i can open a ticket
FYI it works for implicit flow but not authorization code flow

**Comment by jshaughn** [comment]
By the way, if you are unrestricted in your Kiali version you may as well update to the most recent release, 1.30.  See https://kiali.io/documentation/latest/installation-guide/#_version_compatibility for guidance.

**Reply by arctiqaurora** [reply]
i tried upgrading but same error as mentioned: 

> {"error":"failure when retrieving user identity","detail":"request failed (HTTP response status = 401 Unauthorized)"}

error log: 
```kubectl logs kiali-75bb584767-6t272  -n istio-system
2021-03-02T17:13:35Z INF Kiali: Version: v1.30.0, Commit: 9b0d069d1c36965d6fb6902bccc8aaedd1c67fb7

2021-03-02T17:13:35Z INF Using authentication strategy [openid]
2021-03-02T17:13:35Z INF Kiali: Console version: 1.30.0
2021-03-02T17:13:35Z INF Updating base URL in index.html with [/kiali]
2021-03-02T17:13:35Z INF Generating env.js from config
2021-03-02T17:13:35Z INF Server endpoint will start at [:20001/kiali]
2021-03-02T17:13:35Z INF Server endpoint will serve static content from [/opt/kiali/console]
2021-03-02T17:13:35Z INF Starting Metrics Server on [:9090]
2021-03-02T17:17:25Z INF Not handling OpenId code flow authentication: No nonce code present. Login window timed out.
2021-03-02T17:39:14Z INF Not handling OpenId code flow authentication: No nonce code present. Login window timed out.```
